diff -rupN lttoolbox.orig/lttoolbox/fst_processor.cc lttoolbox/lttoolbox/fst_processor.cc
--- lttoolbox.orig/lttoolbox/fst_processor.cc	2009-09-09 16:13:01.000000000 +0200
+++ lttoolbox/lttoolbox/fst_processor.cc	2009-09-09 16:19:10.000000000 +0200
@@ -46,10 +46,18 @@ FSTProcessor::FSTProcessor()
   dictionaryCase = false;
   nullFlush = false;
   nullFlushGeneration = false;
+
+  pool = new Pool<vector<int> >(4, vector<int>(50));
+
+  initial_state = new State(pool);
+  current_state = new State(pool);
 }
 
 FSTProcessor::~FSTProcessor()
 {
+  delete current_state;
+  delete initial_state;
+  delete pool;
 }
 
 void
@@ -419,7 +427,7 @@ FSTProcessor::calcInitial()
     root.addTransition(0, 0, it->second.getInitial());
   }
 
-  initial_state.init(&root);
+  initial_state->init(&root);
 }
 
 bool
@@ -636,7 +644,7 @@ FSTProcessor::analysis(FILE *input, FILE
   bool last_incond = false;
   bool last_postblank = false;
   bool last_preblank = false;
-  State current_state = initial_state;
+  State current_state = *initial_state;
   wstring lf = L"";
   wstring sf = L"";
   int last = 0;
@@ -817,7 +825,7 @@ FSTProcessor::analysis(FILE *input, FILE
         input_buffer.back(1);
       }
 	
-      current_state = initial_state;
+      current_state = *initial_state;
       lf = L"";
       sf = L"";
       last_incond = false;
@@ -900,7 +908,7 @@ FSTProcessor::transliteration_wrapper_nu
 void
 FSTProcessor::tm_analysis(FILE *input, FILE *output)
 {
-  State current_state = initial_state;
+  State current_state = *initial_state;
   wstring lf = L"";
   wstring sf = L"";
   int last = 0;
@@ -1043,7 +1051,7 @@ FSTProcessor::tm_analysis(FILE *input, F
         input_buffer.back(1);
       }
 	
-      current_state = initial_state;
+      current_state = *initial_state;
       lf = L"";
       sf = L"";
     }
@@ -1062,7 +1070,7 @@ FSTProcessor::generation(FILE *input, FI
     generation_wrapper_null_flush(input, output, mode);
   }
 
-  State current_state = initial_state;
+  State current_state = *initial_state;
   wstring sf = L"";
  
   outOfWord = false;
@@ -1134,7 +1142,7 @@ FSTProcessor::generation(FILE *input, FI
         }
       }
   
-      current_state = initial_state;
+      current_state = *initial_state;
       sf = L"";
     }
     else if(iswspace(val) && sf.size() == 0)
@@ -1172,7 +1180,7 @@ FSTProcessor::postgeneration(FILE *input
   }
 
   bool skip_mode = true;
-  State current_state = initial_state;
+  State current_state = *initial_state;
   wstring lf = L"";
   wstring sf = L"";
   int last = 0;
@@ -1318,7 +1326,7 @@ FSTProcessor::postgeneration(FILE *input
 	  }	  
 	}
 
-	current_state = initial_state;
+	current_state = *initial_state;
 	lf = L"";
 	sf = L"";
 	skip_mode = true;
@@ -1338,7 +1346,7 @@ FSTProcessor::transliteration(FILE *inpu
     transliteration_wrapper_null_flush(input, output);
   }
 
-  State current_state = initial_state;
+  State current_state = *initial_state;
   wstring lf = L"";
   wstring sf = L"";
   int last = 0;
@@ -1354,7 +1362,7 @@ FSTProcessor::transliteration(FILE *inpu
       if(!lf.empty()) 
       {
         fputws_unlocked(lf.substr(1).c_str(), output);
-        current_state = initial_state;
+        current_state = *initial_state;
         lf = L"";
         sf = L"";
       }
@@ -1411,7 +1419,7 @@ FSTProcessor::transliteration(FILE *inpu
             fputwc_unlocked(val, output);
           }
         }
-        current_state = initial_state;
+        current_state = *initial_state;
         lf = L"";
         sf = L"";
       }
@@ -1424,7 +1432,7 @@ FSTProcessor::transliteration(FILE *inpu
 wstring
 FSTProcessor::biltrans(wstring const &input_word, bool with_delim)
 {
-  State current_state = initial_state;
+  State current_state = *initial_state;
   wstring result = L"";
   unsigned int start_point = 1;
   unsigned int end_point = input_word.size()-2;
@@ -1587,7 +1595,7 @@ FSTProcessor::biltrans(wstring const &in
 pair<wstring, int>
 FSTProcessor::biltransWithQueue(wstring const &input_word, bool with_delim)
 {
-  State current_state = initial_state;
+  State current_state = *initial_state;
   wstring result = L"";
   unsigned int start_point = 1;
   unsigned int end_point = input_word.size()-2;
@@ -1750,7 +1758,7 @@ FSTProcessor::biltransWithQueue(wstring 
 wstring
 FSTProcessor::biltransWithoutQueue(wstring const &input_word, bool with_delim)
 {
-  State current_state = initial_state;
+  State current_state = *initial_state;
   wstring result = L"";
   unsigned int start_point = 1;
   unsigned int end_point = input_word.size()-2;
@@ -1873,14 +1881,14 @@ FSTProcessor::biltransWithoutQueue(wstri
 bool
 FSTProcessor::valid() const
 {
-  if(initial_state.isFinal(all_finals))
+  if(initial_state->isFinal(all_finals))
   {
     wcerr << L"Error: Invalid dictionary (hint: the left side of an entry is empty)" << endl;
     return false;
   }
   else
   {
-    State s = initial_state;
+    State s = *initial_state;
     s.step(L' ');
     if(s.size() != 0)
     {
@@ -1964,7 +1972,7 @@ FSTProcessor::SAO(FILE *input, FILE *out
 {
   bool last_incond = false;
   bool last_postblank = false;
-  State current_state = initial_state;
+  State current_state = *initial_state;
   wstring lf = L"";
   wstring sf = L"";
   int last = 0;
@@ -2101,7 +2109,7 @@ FSTProcessor::SAO(FILE *input, FILE *out
         input_buffer.back(1);
       }
 	
-      current_state = initial_state;
+      current_state = *initial_state;
       lf = L"";
       sf = L"";
       last_incond = false;
diff -rupN lttoolbox.orig/lttoolbox/fst_processor.h lttoolbox/lttoolbox/fst_processor.h
--- lttoolbox.orig/lttoolbox/fst_processor.h	2009-09-09 16:13:01.000000000 +0200
+++ lttoolbox/lttoolbox/fst_processor.h	2009-09-09 16:19:10.000000000 +0200
@@ -56,15 +56,17 @@ private:
    */
   map<wstring, TransExe, Ltstr> transducers;
 
+  Pool<vector<int> > *pool;
+
   /**
    * Current state of lexical analysis
    */
-  State current_state;
+  State *current_state;
 
   /**
    * Initial state of every token
    */
-  State initial_state;
+  State *initial_state;
 
   /**
    * Set of final states of inconditional sections in the dictionaries
diff -rupN lttoolbox.orig/lttoolbox/state.cc lttoolbox/lttoolbox/state.cc
--- lttoolbox.orig/lttoolbox/state.cc	2009-09-09 16:13:01.000000000 +0200
+++ lttoolbox/lttoolbox/state.cc	2009-09-09 16:19:10.000000000 +0200
@@ -21,10 +21,9 @@
 #include <cstring>
 #include <cwctype>
 
-Pool<vector<int> > State::pool(4, vector<int>(50));
-
-State::State()
+State::State(Pool<vector<int> > *p)
 {
+  pool = p;
 }
  
 State::~State()
@@ -55,7 +54,7 @@ State::destroy()
   // release references
   for(size_t i = 0, limit = state.size(); i != limit; i++)
   {
-    pool.release(state[i].sequence);
+    pool->release(state[i].sequence);
   }
 
   state.clear();
@@ -67,14 +66,15 @@ State::copy(State const &s)
   // release references
   for(size_t i = 0, limit = state.size(); i != limit; i++)
   {
-    pool.release(state[i].sequence);
+    pool->release(state[i].sequence);
   }
 
   state = s.state;
+  pool = s.pool;
 
   for(size_t i = 0, limit = state.size(); i != limit; i++)
   {
-    vector<int> *tmp = pool.get();
+    vector<int> *tmp = pool->get();
     *tmp = *(state[i].sequence);
     state[i].sequence = tmp;
   }
@@ -90,7 +90,7 @@ void
 State::init(Node *initial)
 {
   state.clear();
-  state.push_back(TNodeState(initial,pool.get(),false));
+  state.push_back(TNodeState(initial,pool->get(),false));
   state[0].sequence->clear();
   epsilonClosure();  
 }  
@@ -113,7 +113,7 @@ State::apply(int const input)
     {
       for(int j = 0; j != it->second.size; j++)
       {
-        vector<int> *new_v = pool.get();
+        vector<int> *new_v = pool->get();
         *new_v = *(state[i].sequence);
         if(it->first != 0)
         {
@@ -122,7 +122,7 @@ State::apply(int const input)
         new_state.push_back(TNodeState(it->second.dest[j], new_v, state[i].dirty||false));
       }
     }
-    pool.release(state[i].sequence);
+    pool->release(state[i].sequence);
   }
   
   state = new_state;
@@ -147,7 +147,7 @@ State::apply(int const input, int const 
     {
       for(int j = 0; j != it->second.size; j++)
       {
-        vector<int> *new_v = pool.get();
+        vector<int> *new_v = pool->get();
         *new_v = *(state[i].sequence);
         if(it->first != 0)
         {
@@ -161,7 +161,7 @@ State::apply(int const input, int const 
     {
       for(int j = 0; j != it->second.size; j++)
       {
-        vector<int> *new_v = pool.get();
+        vector<int> *new_v = pool->get();
         *new_v = *(state[i].sequence);
         if(it->first != 0)
         {
@@ -170,7 +170,7 @@ State::apply(int const input, int const 
         new_state.push_back(TNodeState(it->second.dest[j], new_v, true));
       }
     }
-    pool.release(state[i].sequence);
+    pool->release(state[i].sequence);
   }
 
   state = new_state;
@@ -187,7 +187,7 @@ State::epsilonClosure()
     {
       for(int j = 0 ; j != it2->second.size; j++)
       {
-        vector<int> *tmp = pool.get();
+        vector<int> *tmp = pool->get();
         *tmp = *(state[i].sequence);
         if(it2->second.out_tag[j] != 0)
         {
diff -rupN lttoolbox.orig/lttoolbox/state.h lttoolbox/lttoolbox/state.h
--- lttoolbox.orig/lttoolbox/state.h	2009-09-09 16:13:01.000000000 +0200
+++ lttoolbox/lttoolbox/state.h	2009-09-09 16:19:10.000000000 +0200
@@ -60,7 +60,7 @@ private:
   /**
    * Pool of wchar_t vectors, for efficience (static class)
    */
-  static Pool<vector<int> > pool;  
+  Pool<vector<int> > *pool;  
 
   /**
    * Copy function
@@ -96,7 +96,7 @@ public:
   /**
    * Constructor
    */
-  State();
+  State(Pool<vector<int> > *);
 
   /**
    * Destructor
